### PR TITLE
Clarify three roles of consumers of the PMIx interface

### PR DESCRIPTION
**Pull request by jjhursey**
* Clarify the three roles of consumers of the PMIx interface (clients / servers / tools)
 * Part of the implementation agnostic/client separation working group effort
 * Ref Issue #175 and old PR #230

Signed-off-by: Joshua Hursey <jhursey@us.ibm.com>

### CODE REVIEW

**Comment by jjhursey** [comment]
@dsolt I'd like the IAWG to take a look at this before putting it forward. Maybe we can do so next week.
I'll leave it in draft/WIP until the WG gives it 👍

File: Chap_Introduction.tex [hunk 1]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
...
+If a process initializes as either a \emph{server} or a \emph{tool} that process may also access all of the \emph{client} \acp{API}.
+
+A process operating as a \refterm{client} is started (directly or indirectly, for example, by an intermediate script) by the \ac{RM} and is connected to the \ac{PMIx} server instance within that \ac{RM} when the client calls the client \ac{PMIx} initialization routine.
+A process operating as a \declareterm{server} is responsible for starting client processes and coordinating with other server and tool processes in the same \ac{PMIx} universe.
+Often processes operating as a \emph{server} are part of the \acf{RM} infrastructure.
+A process operating as a \declareterm{tool} will connect to a \ac{PMIx} \emph{server} to interact with the processes in the \ac{PMIx} universe.
+An example of a \emph{tool} process is a parallel debugger that will connect to the server to assist with attaching to a set of client processes.
+
+\ac{PMIx} serves as a conduit between processes acting in these three different roles.
+As such, an \ac{API} is often described in how it interacts with processes operating in other roles in the \ac{PMIx} universe.
```
[the remark this answers is not in the document]
Comment (jjhursey) [reply Chap_Introduction.tex]: Good catch. I just pushed with that fix.

**Comment by jjhursey** [comment]
@dsolt Let's add this to the working group agenda to review in the next week or two. We reviewed it last year, so I'm hoping it won't take too long.

File: Chap_Introduction.tex [hunk 2]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
 }
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{PMIx Roles}
+
+The role of a \ac{PMIx} process in the \ac{PMIx} universe is grouped into one of three categories based on how it operates in the \ac{PMIx} environment namely as a \emph{client}, \emph{server}, or \emph{tool}.
+As a result, there are three corresponding sets of initialization and finalization functions.
+If a process initializes as either a \emph{server} or a \emph{tool} that process may also access all of the \emph{client} \acp{API}.
```
Comment (jjhursey) [inline Chap_Introduction.tex:207]:
```suggestion
If a process initializes as either a \emph{server} or a \emph{tool} that process is also considered a \emph{client}.
```

File: Chap_Introduction.tex [hunk 3]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
 }
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{PMIx Roles}
+
+The role of a \ac{PMIx} process in the \ac{PMIx} universe is grouped into one of three categories based on how it operates in the \ac{PMIx} environment namely as a \emph{client}, \emph{server}, or \emph{tool}.
+As a result, there are three corresponding sets of initialization and finalization functions.
```
Comment (jjhursey) [inline Chap_Introduction.tex:206]:
```suggestion
As a result, there are three corresponding sets of initialization and finalization functions and three groupings of \acp{API}.
```

File: Chap_Introduction.tex [hunk 4]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
...
+If a process initializes as either a \emph{server} or a \emph{tool} that process may also access all of the \emph{client} \acp{API}.
+
+A process operating as a \refterm{client} is started (directly or indirectly, for example, by an intermediate script) by the \ac{RM} and is connected to the \ac{PMIx} server instance within that \ac{RM} when the client calls the client \ac{PMIx} initialization routine.
```
Comment (jjhursey) [inline Chap_Introduction.tex:209]:
```suggestion
A process operating as a \refterm{client} is started directly or indirectly (for example, by an intermediate script) by the \ac{RM} and is connected to the \ac{PMIx} server instance within that \ac{RM} when the client calls the client \ac{PMIx} initialization routine.
```

File: Chap_Introduction.tex [hunk 5]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
...
+A process operating as a \declareterm{server} is responsible for starting client processes and coordinating with other server and tool processes in the same \ac{PMIx} universe.
+Often processes operating as a \emph{server} are part of the \acf{RM} infrastructure.
+A process operating as a \declareterm{tool} will connect to a \ac{PMIx} \emph{server} to interact with the processes in the \ac{PMIx} universe.
```
Comment (jjhursey) [inline Chap_Introduction.tex:212]:
```suggestion
A process operating as a \declareterm{tool} is started independently (e.g., via fork/exec) or by the \ac{RM} and will connect to a \ac{PMIx} \emph{server} to interact with the processes in the \ac{PMIx} universe.
```

File: Chap_Introduction.tex [hunk 6]
```diff
@@ -200,3 +200,34 @@ \subsection{Attributes in PMIx}
...
+A \ac{PMIx} implementation that supports all three sets of the \ac{API} role groupings is said to be \emph{fully \ac{PMIx} standard compliant}.
+These \emph{fully \ac{PMIx} standard compliant} implementations have the advantage of being able to support a broad set of \ac{PMIx} consumers in the different roles.
+
+Alternatively, a \ac{PMIx} implementation may choose to support fewer than all three sets of the \ac{API} role groupings.
+A \ac{PMIx} implementation that supports only the \emph{client} \acp{API} is said to be \emph{client-only \ac{PMIx} standard compliant}.
+Similarly, a \ac{PMIx} implementation that only supports the \emph{client} and \emph{tool} \acp{API} is said to be \emph{client-and-tool-only \ac{PMIx} standard compliant}.
+Finally, a \ac{PMIx} implementation that only supports the \emph{client} and \emph{server} \acp{API} is said to be \emph{client-and-server-only \ac{PMIx} standard compliant}.
+
+Note that it would not make sense for an implementation to exclude the \emph{client} interfaces from their implementation since they are also used by the \emph{server} and \emph{tool} roles.
+\adviceimplend
```
Comment (jjhursey) [inline Chap_Introduction.tex:231]: Invert the previous 3 paragraphs. Add "minimal set of the required functionality for PMIx compliance is the client interfaces". From there you can add 'tool' or 'server'. If you have all three then you are `full-role compliant`

Split the advice into a separate commit from the normative text.

Comment (jjhursey) [reply Chap_Introduction.tex:231]: Fixed in latest push.

**Comment by jjhursey** [comment]
I pushed the changes requested from the WG meeting on April 5, 2021

**Comment by jjhursey** [comment]
Notes from PMIx ASC Meeting April 8, 2021:
 * The Client role definition needs to allow for singletons (not started by an RM)
 * Naming:
   * Remove the `-only` suffix
   * Consider moving towards a list and removing the `-and-`. For example `client-role,tool-role \ac{PMIx} standard compliant`
 * Use case compliance can be a separate qualifier on the PMIx and RM requirements.

**Comment by jjhursey** [comment]
Please use emoji reactions ON THIS COMMENT to indicate your position on this proposal.

* You do not need to vote on every proposal
* If you have no opinion, don't vote - that is also useful data
* If you've already commented on this issue, please still vote so
      we know your current thoughts
* Not all proposals solve exactly the same problem, so we may end
      up accepting proposals that appear to have some overlap
This is not a binding majority-rule vote, but it will be a very
significant input into the corresponding ASC decision.

Here are the meanings for the emojis:

* Hooray or Rocket: I support this so strongly that I
      want to be an advocate for it
* Heart: I think this is an ideal solution
* Thumbs up: I'd be happy with this solution
* Confused: I'd rather we not do this, but I can tolerate it
* Thumbs down: I'd be actively unhappy, and may even consider
      other technologies instead
If you want to explain in more detail, feel free to add another
comment, but please also vote on this comment.

File: Chap_Introduction.tex [hunk 7]
```diff
@@ -200,3 +200,36 @@ \subsection{Attributes in PMIx}
...
+A common nomenclature is defined here to aid in identifying levels of conformance of an implementation.
+
+Note that it would not make sense for an implementation to exclude the \emph{client} interfaces from their implementation since they are also used by the \emph{server} and \emph{tool} roles.
+Therefore the \emph{client} interfaces represent the minimal set of required functionality for \ac{PMIx} compliance.
+
+A \ac{PMIx} implementation that supports only the \emph{client} \acp{API} is said to be \emph{client-role \ac{PMIx} standard compliant}.
+Similarly, a \ac{PMIx} implementation that only supports the \emph{client} and \emph{tool} \acp{API} is said to be \emph{client-role and tool-role \ac{PMIx} standard compliant}.
+Finally, a \ac{PMIx} implementation that only supports the \emph{client} and \emph{server} \acp{API} is said to be \emph{client-role and server-role \ac{PMIx} standard compliant}.
+
+A \ac{PMIx} implementation that supports all three sets of the \ac{API} role groupings is said to be \emph{all-role \ac{PMIx} standard compliant}.
```
Comment (jjhursey) [inline Chap_Introduction.tex:231]:
```suggestion
A \ac{PMIx} implementation that supports all three sets of the \ac{API} role groupings is said to be \emph{client-role, server-role, and tool-role \ac{PMIx} standard compliant}.
```

**Comment by kathrynmohror** [comment]
@jjhursey This PR passed its first vote on the condition that you apply the changes in the comments that contain the text 'revision exception'

**Comment by jjhursey** [comment]
* Revision exception changes have now been committed to this branch.
 * I'll bring this for second vote in the 3Q 2021 meeting

**Comment by jjhursey** [comment]
PMIx ASC 2Q 2021: May 11 & May 13
 * Passed first vote with Revision Exception: 7 yes / 0 no / 1 abstain
   - Failed without the Revision Exception): 1 yes / 4 no / 3 abstain
 * Eligible for second vote 3Q 2021 ASC Meeting
 * I confirmed that the Revision Exception has been applied.

**Comment by jjhursey** [comment]
PMIx ASC 3Q 2021
 * Passed the second vote: 7 yes / 0 no / 2 abstain
 * Will be merged into the master branch targeting PMIx v5.0

**Comment by jjhursey** [comment]
I rebased and merged this PR. It'll be in the v5 version of the document.
